File: Amazon/Onsite/C++/Design Search Autocomplete System.cpp
class Data{
public:
    int count;
    string word;
    Data(string _word, int _count) : word( _word), count( _count) {}
    bool operator<( const Data& other) const
    { return count == other.count ? word > other.word : count < other.count; }
};

typedef priority_queue<Data, vector<Data>>  MaxHeap;



class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isword;
    unordered_map<string, int> freq;
      
    TrieNode() {
        isword = false;
    }
};

class Trie {
public:
    Trie(){
        root = new TrieNode();
    }
     
    void insert(const string& word, int freq) {
        TrieNode* curr = root;
        for(auto& c : word) {
            
            if(curr->children.find(c) == curr->children.end())
                curr->children[c] = new TrieNode();
            
            curr = curr->children[c];
            curr->freq[word]+=freq;
        }
        
        curr->isword = true;
    }
    
    vector<string> search(const string& prefix) {
        vector<string> res;
        TrieNode* curr = root;
        for(auto& c : prefix) {
            
            if(curr->children.find(c) == curr->children.end()) return res;
            curr = curr->children[c];
        }
        
        MaxHeap maxhp;
        for(auto& word : curr->freq)
            maxhp.emplace(word.first, word.second);
        
        for(int i = 0; !maxhp.empty() && i < 3; ++i)
        {   res.push_back(maxhp.top().word); maxhp.pop(); }
        
        return res;
    }
    
private:
    TrieNode* root;
};


class AutocompleteSystem {
public:
    AutocompleteSystem(vector<string>& sentences, vector<int>& times) {
        trie = new Trie();
        for(int i = 0; i < sentences.size(); ++i)
            trie->insert(sentences[i], times[i]);
    }
    
    vector<string> input(char c) {
        if(c == '#') {
            trie->insert(prefix, 1);
            prefix = "";
            res.clear();
            return res;
        }
        
        prefix+=c;
        auto res = trie->search(prefix);
        return res;
    }
private:
    Trie* trie;
    string prefix="";
    vector<string> res;
};

File: Amazon/Onsite/Python/Design Search Auto-complete System.py
class Node:
    def __init__(self):
        self.children = {}
        self.isword = False
        self.top3 = collections.defaultdict(list) # store index of top3 words []
class Trie:
    def __init__(self):
        self.root = Node()
        
    def add(self, word, wordlist, worddict):
        cur = self.root
        idx = worddict[word][0] # idx
        for w in word:
            if w not in cur.children:
                cur.children[w] = Node()
                cur.top3[w].append(idx)
            else:
                if idx not in cur.top3[w]: cur.top3[w].append(idx)
                cur.top3[w].sort(key = lambda x: (-worddict[wordlist[x]][1], wordlist[x])) # -freq, word
                if len(cur.top3[w]) > 3: cur.top3[w].pop()
            cur = cur.children[w]
        cur.isword = True
        
class AutocompleteSystem(object):
    def __init__(self, sentences, times):
        self.t = Trie()
        n = len(sentences)
        self.wordlist = sentences # list, wordlist[idx] = word
        self.worddict = {}        # dict, worddict[word] = [idx, freq]
        for idx in range(n):
            self.worddict[sentences[idx]] = [idx, times[idx]]
            self.t.add(sentences[idx], self.wordlist, self.worddict)
        self.cur = self.t.root
        self.newword = []
        
    def input(self, c):
        if c == '#': 
            nw = ''.join(self.newword)
            if nw not in self.worddict:
                self.wordlist.append(nw)
                self.worddict[nw] = [len(self.wordlist)-1, 1]
            else:
                self.worddict[nw][1] += 1 # update freq
            self.t.add(nw, self.wordlist, self.worddict)
            # reset
            self.cur = self.t.root
            self.newword = []
            return []
        
        self.newword.append(c) # keep recording
        if self.cur and c in self.cur.children:
            res = [self.wordlist[idx] for idx in self.cur.top3[c]]
            self.cur = self.cur.children[c]
            return res
        else:
            self.cur = None # stop search
            return []